slay/truss_adapter/deploy.py

```diff
@@ -0,0 +1,325 @@
+            "to secrets - no need to manually add it."
+        )
+    config.model_cache.models = assets.cached
+    # Metadata.
+    slay_metadata: definitions.TrussMetadata = definitions.TrussMetadata(
+        user_config=slay_config.user_config,
+        stub_cls_to_url=stub_cls_to_url,
+    )
+    config.model_metadata[definitions.TRUSS_CONFIG_SLAY_KEY] = slay_metadata.dict()
+    return config
+
+
+def make_truss(
+    processor_dir: pathlib.Path,
+    workflow_root: pathlib.Path,
+    slay_config: definitions.Config,
+    model_name: str,
+    stub_cls_to_url: Mapping[str, str],
+    maybe_stub_file: Optional[pathlib.Path],
+) -> pathlib.Path:
+    truss_dir = processor_dir / _TRUSS_DIR
+    truss_dir.mkdir(exist_ok=True)
+    config = _make_truss_config(truss_dir, slay_config, stub_cls_to_url, model_name)
+
+    config.write_to_yaml_file(
+        truss_dir / serving_image_builder.CONFIG_FILE, verbose=False
+    )
+
+    # Copy other sources.
+    model_dir = truss_dir / truss_config.DEFAULT_MODEL_MODULE_DIR
+    model_dir.mkdir(parents=True, exist_ok=True)
+    shutil.copy(
+        processor_dir / f"{definitions.PROCESSOR_MODULE}.py",
+        model_dir / _MODEL_CLASS_FILENAME,
+    )
+
+    pkg_dir = truss_dir / truss_config.DEFAULT_BUNDLED_PACKAGES_DIR
+    pkg_dir.mkdir(parents=True, exist_ok=True)
+    if maybe_stub_file is not None:
+        shutil.copy(maybe_stub_file, pkg_dir)
+    # TODO This assume all imports are absolute w.r.t workflow root (or site-packages).
+    # Also: apparently packages need an `__init__`, or crash.
+    _copy_python_source_files(workflow_root, pkg_dir / pkg_dir)
+
+    # TODO Truss package contains this from `{ include = "slay", from = "." }`
+    # pyproject.toml. But for quick dev loop just copy from local.
+    shutil.copytree(
+        os.path.dirname(slay.__file__),
+        pkg_dir / "slay",
+        dirs_exist_ok=True,
+    )
+    return truss_dir
+
+
+def get_api_key_from_truss_config() -> str:
+    return remote_factory.load_config().get("baseten", "api_key")
+
+
+class _BasetenEnv(enum.Enum):
+    LOCAL = enum.auto()
+    STAGING = enum.auto()
+    PROD = enum.auto()
+    DEV = enum.auto()
+
+
+def _infer_env(baseten_url: str) -> _BasetenEnv:
+    if baseten_url in {"localhost", "127.0.0.1", "0.0.0.0"}:
+        return _BasetenEnv.LOCAL
+
+    if "staging" in baseten_url:
+        return _BasetenEnv.STAGING
+
+    if "dev" in baseten_url:
+        return _BasetenEnv.DEV
+
+    return _BasetenEnv.PROD
+
+
+def _model_url(baseten_env: _BasetenEnv, model_id: str) -> str:
+    if baseten_env == _BasetenEnv.LOCAL:
+        return f"http://localhost:8000/models/{model_id}"
+
+    if baseten_env == _BasetenEnv.STAGING:
+        return f"https://app.staging.baseten.co/models/{model_id}"
+
+    if baseten_env == _BasetenEnv.DEV:
+        return f"https://app.dev.baseten.co/models/{model_id}"
+
+    return f"https://model-{model_id}.api.baseten.co/production"
```

Comment:
nit: this should be consistent with the above. Or you can change the above to `model-{model_ind}.api.{env.lower()}.baseten.co` with the exceptioon of local

Reply:
I'll add a TODO. I actually want to change all of this here quite a bit, using the REST API instead. But it would take me some time to test/debug, and I know that the current state works - so leaving it in this PR.